migrate.py

```diff
@@ -0,0 +1,217 @@
+        self._post(dataset_name)
+
+    def _find_newest_snapshot_in_restic(self, snapshots_in_restic: Dict[str, str]):
+        snapshots = list(snapshots_in_restic.keys())
+        snapshots.sort()
+        return snapshots[-1]
+
+    def _find_next_snapshot(self, dataset_name: str, snapshots: List[str], newest_snapshot_in_restic: str) -> Optional[str]:
+        oldest_found = None
+        for snapshot in snapshots:
+            if snapshot <= newest_snapshot_in_restic:
+                continue
+            if oldest_found is not None and snapshot > oldest_found:
+                # snapshot is newer than another we found
+                continue
+            if self._get_zfs_snapshot_size(dataset_name, snapshot) == 0:
+                print(F"Skipping snapshot {dataset_name}@{snapshot} because of zero size.")
+                continue
+            oldest_found = snapshot
+        return oldest_found
+
+    def _backup_next_snapshot_from_dataset(self, dataset_name) -> bool:
+        restic_repo, _ = self._get_repo_name_and_path(dataset_name)
+
+        snapshots = self._get_dataset_snapshots(dataset_name)
+        snapshots_in_restic = self._get_snapshots_in_restic(restic_repo)
+        parent_restic_snapshot = None
+        newest_snapshot_in_restic = "00000000"
+        if len(snapshots_in_restic) > 0:
+            newest_snapshot_in_restic = self._find_newest_snapshot_in_restic(snapshots_in_restic)
+            parent_restic_snapshot = snapshots_in_restic[newest_snapshot_in_restic]
+        snapshot = self._find_next_snapshot(dataset_name, snapshots, newest_snapshot_in_restic)
+        if snapshot is None:
+            print(f"No further snapshots need to backuped for {dataset_name}.")
+            return False
+        self._backup_single_snapshot(dataset_name, snapshot, parent_restic_snapshot)
+        return True
+
+    def backup_next_snapshot_from_dataset(self, dataset_name):
+        self._pre(dataset_name)
+        self._backup_next_snapshot_from_dataset(dataset_name)
+        self._post(dataset_name)
+
+    def _backup_dataset(self, dataset_name: str):
+        while self._backup_next_snapshot_from_dataset(dataset_name):
+            pass
+
+    def backup_dataset(self, dataset_name: str):
+        self._pre(dataset_name)
+        self._backup_dataset(dataset_name)
+        self._post(dataset_name)
+
+
+def main():
+    if os.geteuid() != 0:
+        print("Please run as root.")
+        exit(1)
+    parser = argparse.ArgumentParser(description='Migrate zfs backups to restic.')
+    parser.add_argument('-r', '--restic-repo-prefix', required=True,
+                        help='The prefix used for the restic repo. It is appended with the dataset name.')
+    parser.add_argument('-c', '--zfs-dataset-common-prefix', default="",
+                        help='The prefix which should be removed from each dataset name for use in the restic repo. Eg. backup01')
+    parser.add_argument('-p', '--restic-password-file', required=True,
```

Comment:
No need to require this, I think. You could just export `RESTIC_REPOSITORY_FILE` which is recognized by restic.